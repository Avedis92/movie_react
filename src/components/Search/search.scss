@import "../sasscommonmixxins/flex.scss";
@import "../sasscommonmixxins/button.scss";
@import "../sasscommonmixxins/mediaqueries.scss";

.searchElement {
  background-size: 100% 100%;
  background-repeat: no-repeat;
  background-position-x: center;
  width: 100%;
  height: 600px;
  color: white;

  .searchFooter {
    margin-left: 5%;
    margin-bottom: 5%;
    padding-top: 15%;
    height: 50%;
    @include flexElements(flex, column, none, space-between);
    .titleAndGenresAndOverview {
      margin-right: 5%;
      @include flexElements(flex, row, none, space-between);
      .headerandGenreList {
        width: 40%;
        .searchedMovieGenreList {
          h2 {
            display: inline;
          }
        }
        @include mobile {
          width: 100%;
          h1 {
            font-size: 26px;
            display: inline;
          }
          h2 {
            font-size: 20px;
            margin-bottom: 4%;
          }
        }
      }

      .searchedMovieOverview {
        width: 25%;
        background-color: rgba(0, 0, 0, 0.541);

        p {
          padding: 10px;
        }
        @include mobile {
          width: 100%;
          margin-bottom: 4%;
        }
      }
      @include mobile {
        @include flexElements(flex, column, none, space-between);
      }
    }
    .searchedMovieExtraDetails {
      @include flexElements(flex, row, center, space-between);
      height: 15%;
      .searchedMovieRatings {
        @include flexElements(flex, row, center, space-between);
        width: 20%;
        .searchedMovieVoteAverage {
          border: 2px solid rgb(27, 186, 235);
          width: 15%;
          text-align: center;
          border-radius: 5px;
          p {
            margin-top: 3%;
            margin-bottom: 3%;
          }
          @include mobile {
            width: 10vw;
          }
        }
        @include mobile {
          width: 40vw;
        }
      }
      .searchedMovieExtraInfoButtons {
        @include flexElements(flex, row, none, space-evenly);
        width: 25%;
        height: 100%;
        margin-right: 5%;
        @include mobile {
          width: 45vw;
          @include flexElements(flex, row, none, space-between);
        }
      }
    }
  }
  @include mobile {
    height: 450px;
  }
}
