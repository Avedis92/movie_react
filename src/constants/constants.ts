export const apiKey = '0600befd930f4be78f3d1123f393affc';
export const baseUrl = 'https://api.themoviedb.org/3/';
export const trendingMovieUrl = `${baseUrl}trending/movie/week?api_key=${apiKey}&page=`;
export const topRatedMovieUrl = `${baseUrl}movie/top_rated?api_key=${apiKey}&page=`;
export const upcomingMovieUrl = `${baseUrl}movie/upcoming?api_key=${apiKey}&page=`;
export const movieGenreListUrl = `${baseUrl}genre/movie/list?api_key=${apiKey}`;
export const searchMovieUrl = `${baseUrl}search/movie?api_key=${apiKey}&query=`;
export const postPathUrl = 'https://image.tmdb.org/t/p/w500';
export const defaultBackgroundImage = 'https://media-assets-05.thedrum.com/cache/images/thedrum-prod/s3-news-tmp-10557-netflixa--default--1280.jpg';
export const pageNotFoundBackgroundImage = 'https://cdn3.vectorstock.com/i/1000x1000/67/12/404-error-page-background-vector-20596712.jpg';
export const movieVideoUrl = `${baseUrl}/movie/`;
export const searchMovieByGenreUrl = `${baseUrl}discover/movie?api_key=${apiKey}&with_genres=`;
export const defaultPageNumber = 1;
export const defaultMovieType = 'trendings';
