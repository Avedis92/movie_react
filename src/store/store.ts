import { configureStore } from '@reduxjs/toolkit';
import genreReducer from './slices/genre';

export const store = configureStore({
  reducer: {
    genre: genreReducer,
  },
});

export type RootState=ReturnType<typeof store.getState>;
export type MovieDispatch=typeof store.dispatch;
