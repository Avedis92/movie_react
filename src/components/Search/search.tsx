import { useState } from 'react';
import { useHistory } from 'react-router-dom';
import { useSelector } from 'react-redux';
import ModalVideo from 'react-modal-video';
import { RootState } from '../../store/store';
import 'react-modal-video/scss/modal-video.scss';
import './search.scss';
import RatingStars from '../Rating/rating';
import Loader from '../Loader/loader';
import TrailerPlayer from '../TrailerPlayer/trailerplayer';
import InfoButtons from '../InfoButtons/infobuttons';
import * as constants from '../../constants/constants';

interface SearchMovie{
  movieId?:number,
  originalTitle?:string,
  movieBackgroundImage?:string,
  genreList?:Array<number>,
  movieOverview?:string,
  movieVoteAverage?:number,
  searchMovieLoading:boolean;
}

function Search({
  movieId, originalTitle, movieBackgroundImage, genreList, movieOverview, movieVoteAverage, searchMovieLoading,
}:SearchMovie) {
  const history = useHistory();
  const allgenre = useSelector((state:RootState) => state.genre);
  const [isOpen, setIsOpen] = useState(false);
  const [trailerKey, setTrailerKey] = useState('');
  const getGenres = (genreId:number) => {
    if (allgenre) {
      return allgenre.entities.find((genre:Genre) => genre.id === genreId);
    }
    return [];
  };
  let myGenres:Array<Genre> = [];
  myGenres = genreList && genreList.length > 0 ? genreList.map(
    (myid) => getGenres(myid),
  ) : ['N/A'];
  return (
    <>
      {searchMovieLoading === false ? (
        <div
          className="searchElement"
          style={{
            backgroundImage:
      `url(${movieBackgroundImage !== `${constants.postPathUrl}null` ? movieBackgroundImage : constants.defaultBackgroundImage}`,
          }}
        >
          {isOpen ? (
            <ModalVideo
              channel="youtube"
              isOpen={isOpen}
              videoId={trailerKey}
              onClose={() => setIsOpen(false)}
            />
          ) : null}
          <div className="searchFooter">
            {movieId ? (
              <>
                <div className="titleAndGenresAndOverview">
                  <div className="headerandGenreList">
                    <h1 data-testid="searchedTitle">{originalTitle}</h1>
                    <div className="searchedMovieGenreList">
                      { myGenres ? myGenres.map((myMovieGenre, index) => (
                        <h2 key={index}>
                          {myMovieGenre && myMovieGenre.name}
                          {' '}
                          {' '}
                        </h2>
                      )) : null}
                    </div>
                  </div>
                  <div className="searchedMovieOverview">
                    <p>{movieOverview}</p>
                  </div>
                </div>
                <div className="searchedMovieExtraDetails">
                  <div className="searchedMovieRatings">
                    <RatingStars movieRating={movieVoteAverage} />
                    <div className="searchedMovieVoteAverage">
                      <p>{movieVoteAverage}</p>
                    </div>
                  </div>
                  <div className="searchedMovieExtraInfoButtons">
                    <TrailerPlayer
                      movieId={movieId}
                      buttonType="watchNowButton"
                      openMovieTrailers={(key) => {
                        setTrailerKey(key);
                        setIsOpen(true);
                      }}
                    />
                    <InfoButtons
                      moreInfo={() => history.push(`/moviedetailpage?movieId=${movieId}`)}
                    />
                  </div>
                </div>
              </>
            ) : null}

          </div>

        </div>
      ) : <Loader />}
    </>
  );
}
export default Search;
