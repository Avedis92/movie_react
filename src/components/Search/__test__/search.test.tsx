import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { store } from '../../../store/store';
import Search from '../search';
import * as constants from '../../../constants/constants';
import '@testing-library/jest-dom/extend-expect';

const genres = {
  entities: [
    {
      id: 28,
      name: 'Action',
    },
    {
      id: 12,
      name: 'Adventure',
    },
    {
      id: 16,
      name: 'Animation',
    },
    {
      id: 35,
      name: 'Comedy',
    },
    {
      id: 80,
      name: 'Crime',
    },
    {
      id: 99,
      name: 'Documentary',
    },
    {
      id: 18,
      name: 'Drama',
    },
    {
      id: 10751,
      name: 'Family',
    },
    {
      id: 14,
      name: 'Fantasy',
    },
    {
      id: 36,
      name: 'History',
    },
    {
      id: 27,
      name: 'Horror',
    },
    {
      id: 10402,
      name: 'Music',
    },
    {
      id: 9648,
      name: 'Mystery',
    },
    {
      id: 10749,
      name: 'Romance',
    },
    {
      id: 878,
      name: 'Science Fiction',
    },
    {
      id: 10770,
      name: 'TV Movie',
    },
    {
      id: 53,
      name: 'Thriller',
    },
    {
      id: 10752,
      name: 'War',
    },
    {
      id: 37,
      name: 'Western',
    },
  ],
};

describe('Create tests on Search component', () => {
  jest.mock('react-redux', () => ({
    useSelector: jest.fn().mockReturnValue(genres),
  }));
  test('Make a snapshot test', () => {
    const {asFragment} = render(
      <Provider store={store}>
        <Search
          movieId={297761}
          genreList={[14, 28, 80]}
          originalTitle="Suicide Squad"
          movieVoteAverage={5.91}
          movieOverview="From DC Comics comes the Suicide Squad, an antihero team of incarcerated supervillains who act as deniable assets for the United States government, undertaking high-risk black ops missions in exchange for commuted prison sentences."
          movieBackgroundImage={`${constants.postPathUrl}/ndlQ2Cuc3cjTL7lTynw6I4boP4S.jpg`}
          searchMovieLoading={false}
        />
      </Provider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
