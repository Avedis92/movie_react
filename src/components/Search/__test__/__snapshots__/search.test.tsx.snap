// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create tests on Search component Make a snapshot test 1`] = `
<div
  className="searchElement"
  style={
    Object {
      "backgroundImage": "url(https://image.tmdb.org/t/p/w500/ndlQ2Cuc3cjTL7lTynw6I4boP4S.jpg",
    }
  }
>
  <div
    className="searchFooter"
  >
    <div
      className="titleAndGenresAndOverview"
    >
      <div
        className="headerandGenreList"
      >
        <h1
          data-testid="searchedTitle"
        >
          Suicide Squad
        </h1>
        <div
          className="searchedMovieGenreList"
        >
          <h2>
             
             
          </h2>
          <h2>
             
             
          </h2>
          <h2>
             
             
          </h2>
        </div>
      </div>
      <div
        className="searchedMovieOverview"
      >
        <p>
          From DC Comics comes the Suicide Squad, an antihero team of incarcerated supervillains who act as deniable assets for the United States government, undertaking high-risk black ops missions in exchange for commuted prison sentences.
        </p>
      </div>
    </div>
    <div
      className="searchedMovieExtraDetails"
    >
      <div
        className="searchedMovieRatings"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": "rgb(27, 186, 235)",
            }
          }
          viewBox="0 0 576 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M259.3 17.8L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0z"
          />
        </svg>
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": "rgb(27, 186, 235)",
            }
          }
          viewBox="0 0 576 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M259.3 17.8L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0z"
          />
        </svg>
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": "rgb(27, 186, 235)",
            }
          }
          viewBox="0 0 576 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M259.3 17.8L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0z"
          />
        </svg>
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": "rgb(27, 186, 235)",
            }
          }
          viewBox="0 0 576 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M528.1 171.5L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6zM388.6 312.3l23.7 138.4L288 385.4l-124.3 65.3 23.7-138.4-100.6-98 139-20.2 62.2-126 62.2 126 139 20.2-100.6 98z"
          />
        </svg>
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": "rgb(27, 186, 235)",
            }
          }
          viewBox="0 0 576 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M528.1 171.5L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6zM388.6 312.3l23.7 138.4L288 385.4l-124.3 65.3 23.7-138.4-100.6-98 139-20.2 62.2-126 62.2 126 139 20.2-100.6 98z"
          />
        </svg>
        <div
          className="searchedMovieVoteAverage"
        >
          <p>
            5.91
          </p>
        </div>
      </div>
      <div
        className="searchedMovieExtraInfoButtons"
      >
        <div
          className="watchNowContainer"
        >
          <button
            id="297761"
            name="watchNowButton"
            onClick={[Function]}
            type="button"
          >
            Watch Now
          </button>
        </div>
        <div
          className="movieInfoButtons"
        >
          <button
            name="moreinfo"
            onClick={[Function]}
            type="button"
          >
            More Info
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
